packages/agent/src/certificate.ts

```diff
@@ -130,28 +129,41 @@ export class Certificate {
     }
   }
 
-  private async _checkDelegation(d?: Delegation): Promise<ArrayBuffer> {
+  private async _checkDelegationAndGetKey(
+    canisterId: Principal,
+    d?: Delegation,
+  ): Promise<ArrayBuffer> {
     if (!d) {
-      if (!this._rootKey) {
```

Comment:
Why removing this logic? `fetchRootKey()` should only be called in local replica. The root key for the mainnet is hardcoded and doesn't require explicit fetching.

Reply:
Thanks for catching this - my fix had made things worse! However, I do think that this is the wrong place for this logic - there's no real need for `Certificate` to know anything about the agent IMO (I ended up staring at this piece of code for a while when first looking it, and obviously got it wrong), it just needs a root key. 

I changed the code now so that the callers use the root key if available. We can probably extract a convenience method on `Agent` to do this.